prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+		if !kerrors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get prowjob %s: %v", request.Name, err)
+		}
+
+		// Objects can be deleted from the API while being in our workqueue
+		return reconcile.Result{}, nil
+	}
+
+	// TODO: Terminal errors for unfixable cases like missing build clusters
+	// and not return an error to prevent requeuing?
+	res, err := r.reconcile(pj)
+	if res == nil {
+		res = &reconcile.Result{}
+	}
+	if err != nil {
+		r.log.WithError(err).Error("Reconciliation failed")
+	}
+	return *res, err
+}
+
+func (r *reconciler) reconcile(pj *prowv1.ProwJob) (*reconcile.Result, error) {
+	// terminateDupes first, as that might reduce cluster load and prevent us
+	// from doing pointless work.
+	if err := r.terminateDupes(pj); err != nil {
+		return nil, fmt.Errorf("terminateDupes failed: %w", err)
+	}
+
+	switch pj.Status.State {
+	case prowv1.PendingState:
+		return nil, r.syncPendingJob(pj)
+	case prowv1.TriggeredState:
+		return r.syncTriggeredJob(pj)
+	case prowv1.AbortedState:
+		return nil, r.syncAbortedJob(pj)
+	}
+
+	return nil, nil
+}
+
+func (r *reconciler) terminateDupes(pj *prowv1.ProwJob) error {
+	pjs := &prowv1.ProwJobList{}
+	if err := r.pjClient.List(r.ctx, pjs, optNotCompletedProwJobs()); err != nil {
+		return fmt.Errorf("failed to list prowjobs: %v", err)
+	}
+
+	return pjutil.TerminateOlderJobs(r.pjClient, r.log, pjs.Items, func(toCancel prowv1.ProwJob) error {
+		client, ok := r.buildClients[pj.ClusterAlias()]
+		if !ok {
+			return fmt.Errorf("no client for cluster %q present", pj.ClusterAlias())
+		}
+		podToDelete := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: r.config().PodNamespace,
+				Name:      toCancel.Name,
+			},
+		}
+		if err := client.Delete(r.ctx, podToDelete); err != nil {
+			return fmt.Errorf("failed to delete pod %s/%s: %w", podToDelete.Namespace, podToDelete.Name, err)
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) syncPendingJob(pj *prowv1.ProwJob) error {
+	prevPJ := pj.DeepCopy()
+
+	pod, podExists, err := r.pod(pj)
+	if err != nil {
+		return err
+	}
+
+	if !podExists {
+		// Pod is missing. This can happen in case the previous pod was deleted manually or by
+		// a rescheduler. Start a new pod.
+		id, pn, err := r.startPod(pj)
+		if err != nil {
+			if !isRequestError(err) {
+				return fmt.Errorf("error starting pod %s: %v", pod.Name, err)
+			}
+			pj.Status.State = prowv1.ErrorState
+			pj.SetComplete()
+			pj.Status.Description = "Job cannot be processed."
+			r.log.WithFields(pjutil.ProwJobFields(pj)).WithError(err).Warning("Unprocessable pod.")
+		} else {
+			pj.Status.BuildID = id
+			pj.Status.PodName = pn
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Pod is missing, starting a new pod")
+		}
+	} else {
+
+		switch pod.Status.Phase {
+		case corev1.PodUnknown:
+			// Pod is in Unknown state. This can happen if there is a problem with
+			// the node. Delete the old pod, we'll start a new one next loop.
```

Comment:
nit: no more loops

Reply:
yeah the comment is misleading, but the pod deletion will trigger us again

Reply:
Updated it